core/shared/src/main/scala/zio/internal/macros/LayerBuilder.scala

```diff
@@ -132,6 +129,79 @@ final case class LayerBuilder[Type, Expr](
     }
   }
 
+  /**
+   * Returns tree representing final `ZLayer[In, Err, Out]`
+   *
+   * ==If [[method]] is not [[ProvideMethod.ProvideSomeShared]]==
+   *   - `In` will be some autogenerated type
+   *   - `Err` will be the union of all errors in the provided layers
+   *   - `Out` will be AND'ed together types from [[target0]]
+   *
+   * ==If [[method]] is [[ProvideMethod.ProvideSomeShared]]==
+   *   - `In` will be some autogenerated type
+   *   - `Err` will be the union of all errors in the provided layers
+   *   - `Out` will be some autogenerated type
+   *
+   * And following relation exists: `In & Out` <:< AND'ed together types from
+   * [[target0]]
+   */
+  def buildAuto: Expr = {
+    assertNoAmbiguity()
+
+    val remainderTypeFactory = (t: Type) => {
+      Some(typeToNode(t))
+    }
+
+    val nodes: List[Node[Type, Expr]] = providedLayerNodes ++ sideEffectNodes
+    val nodesOutputs                  = nodes.flatMap(n => n.outputs)
+    val graph                         = Graph(nodes, typeEquals, remainderTypeFactory)
+
+    // ProvideSomeShared ignores optionalInputs - they are the "shared" part
+    val (requiredInputs, optionalInputs) = target0
+      .partition(t => nodesOutputs.exists(tpe => typeEquals(tpe, t)))
+
+    /**
+     * Build the layer tree. This represents the structure of a successfully
+     * constructed ZLayer that will build the target types. This, of course, may
+     * fail with one or more GraphErrors.
+     */
+    val layerTreeEither: Either[::[GraphError[Type, Expr]], LayerTree[Expr]] =
+      for {
+        inputs <- if (method.isProvideSomeShared) {
+                    graph.buildComplete(requiredInputs)
+                  } else {
+                    graph.buildComplete(requiredInputs ++ optionalInputs)
+                  }
+        sideEffects <- graph.buildNodes(sideEffectNodes)
+      } yield sideEffects ++ inputs
+
+    layerTreeEither match {
+      case Left(buildErrors) =>
+        reportBuildErrors(buildErrors)
+
+      case Right(tree) =>
+        // TODO: Write some tests for previous implementation and prepare new implementation based on them
```

Comment:
Leftover or still needs to be done?

Reply:
The notion of `unused` in auto is a bit complicated - I want to implement it in another MR. For initial implementation this warning is not really necessary,